File: src/com/moneydance/modules/features/ruby/RubyConsole.java
package com.moneydance.modules.features.ruby;

import org.jruby.Ruby;
import org.jruby.RubyObject;
import org.jruby.javasupport.util.RuntimeHelpers;
import org.jruby.runtime.builtin.IRubyObject;
import org.jruby.javasupport.JavaUtil;
import org.jruby.RubyClass;


public class RubyConsole extends RubyObject  {
    private static final Ruby __ruby__ = Ruby.getGlobalRuntime();
    private static final RubyClass __metaclass__;

    static {
        String source = new StringBuilder("require 'jruby'\n" +
            "require 'irb'\n" +
            "require 'irb/completion'\n" +
            "\n" +
            "#java_require 'moneydance_irb'  # Require this file, instead of pasting code into .java\n" +
            "java_package 'com.moneydance.modules.features.ruby'\n" +
            "\n" +
            "# Moneydance IRB Console\n" +
            "class RubyConsole\n" +
            "\n" +
            "  if ARGV.empty?\n" +
            "    # default options, esp. useful for jrubyw\n" +
            "    ARGV << '--readline' << '--prompt' << 'inf-ruby'\n" +
            "  end\n" +
            "\n" +
            "  import java.awt.Color\n" +
            "  import java.awt.Font\n" +
            "  import javax.swing.JFrame\n" +
            "  import java.awt.EventQueue\n" +
            "\n" +
            "  # Dummy static method\n" +
            "  def self.noop\n" +
            "\n" +
            "  end\n" +
            "\n" +
            "# Try to find preferred font family, use otherwise -- err --  otherwise\n" +
            "  def self.find_font(otherwise, style, size, *families)\n" +
            "    avail_families = java.awt.GraphicsEnvironment.local_graphics_environment.available_font_family_names\n" +
            "    fontname = families.find(proc { otherwise }) { |name| avail_families.include? name }\n" +
            "    Font.new(fontname, style, size)\n" +
            "  end\n" +
            "\n" +
            "  text = javax.swing.JTextPane.new\n" +
            "  text.font = find_font('Monospaced', Font::PLAIN, 14, 'Monaco', 'Andale Mono')\n" +
            "  text.margin = java.awt.Insets.new(8, 8, 8, 8)\n" +
            "  text.caret_color = Color.new(0xa4, 0x00, 0x00)\n" +
            "  text.background = Color.new(0xf2, 0xf2, 0xf2)\n" +
            "  text.foreground = Color.new(0xa4, 0x00, 0x00)\n" +
            "\n" +
            "  pane = javax.swing.JScrollPane.new\n" +
            "  pane.viewport_view = text\n" +
            "  frame = JFrame.new('Moneydance Interactive Ruby Console (tab will autocomplete)')\n" +
            "#  frame.default_close_operation = JFrame::EXIT_ON_CLOSE\n" +
            "#  frame.default_close_operation = JFrame::DISPOSE_ON_CLOSE\n" +
            "  frame.default_close_operation = JFrame::HIDE_ON_CLOSE\n" +
            "\n" +
            "#  frame.addWindowListener(WindowAdapter.new {\n" +
            "#public void windowClosing(WindowEvent e) {\n" +
            "#System.exit(0);\n" +
            "#}\n" +
            "#});\n" +
            "  frame.set_size(700, 600)\n" +
            "  frame.content_pane.add(pane)\n" +
            "  tar = org.jruby.demo.TextAreaReadline.new(text,\n" +
            "                                            \" Welcome to the Moneydance IRB Console [#{JRUBY_VERSION}] \\n\\n\")\n" +
            "  JRuby.objectspace = true # useful for code completion\n" +
            "  tar.hook_into_runtime_with_streams(JRuby.runtime)\n" +
            "\n" +
            "  EventQueue.invoke_later proc { frame.visible = true; STDOUT.puts \"SVisible...\"; sleep 1 }\n" +
            "\n" +
            "  IRB.start(__FILE__)\n" +
            "\n" +
            "  STDOUT.puts \"IRB finished...\"\n" +
            "\n" +
            "  def dispose\n" +
            "    STDOUT.puts \"Disposing...\"\n" +
            "    frame.dispose\n" +
            "    STDOUT.puts \"Disposed...\"\n" +
            "  end\n" +
            "end\n" +
            "").toString();
        __ruby__.executeScript(source, "lib/ruby_console.rb");
        RubyClass metaclass = __ruby__.getClass("RubyConsole");
        metaclass.setRubyStaticAllocator(RubyConsole.class);
        if (metaclass == null) throw new NoClassDefFoundError("Could not load Ruby class: RubyConsole");
        __metaclass__ = metaclass;
    }

    /**
     * Standard Ruby object constructor, for construction-from-Ruby purposes.
     * Generally not for user consumption.
     *
     * @param ruby The JRuby instance this object will belong to
     * @param metaclass The RubyClass representing the Ruby class of this object
     */
    private RubyConsole(Ruby ruby, RubyClass metaclass) {
        super(ruby, metaclass);
    }

    /**
     * A static method used by JRuby for allocating instances of this object
     * from Ruby. Generally not for user comsumption.
     *
     * @param ruby The JRuby instance this object will belong to
     * @param metaclass The RubyClass representing the Ruby class of this object
     */
    public static IRubyObject __allocate__(Ruby ruby, RubyClass metaClass) {
        return new RubyConsole(ruby, metaClass);
    }
        
    /**
     * Default constructor. Invokes this(Ruby, RubyClass) with the classloader-static
     * Ruby and RubyClass instances assocated with this class, and then invokes the
     * no-argument 'initialize' method in Ruby.
     *
     * @param ruby The JRuby instance this object will belong to
     * @param metaclass The RubyClass representing the Ruby class of this object
     */
    public RubyConsole() {
        this(__ruby__, __metaclass__);
        RuntimeHelpers.invoke(__ruby__.getCurrentContext(), this, "initialize");
    }

    
    public static Object noop() {

        IRubyObject ruby_result = RuntimeHelpers.invoke(__ruby__.getCurrentContext(), __metaclass__, "noop");
        return (Object)ruby_result.toJava(Object.class);

    }

    
    public static Object find_font(Object otherwise, Object style, Object size, Object families) {
        IRubyObject ruby_args[] = new IRubyObject[4];
        ruby_args[0] = JavaUtil.convertJavaToRuby(__ruby__, otherwise);
        ruby_args[1] = JavaUtil.convertJavaToRuby(__ruby__, style);
        ruby_args[2] = JavaUtil.convertJavaToRuby(__ruby__, size);
        ruby_args[3] = JavaUtil.convertJavaToRuby(__ruby__, families);

        IRubyObject ruby_result = RuntimeHelpers.invoke(__ruby__.getCurrentContext(), __metaclass__, "find_font", ruby_args);
        return (Object)ruby_result.toJava(Object.class);

    }

    
    public Object dispose() {

        IRubyObject ruby_result = RuntimeHelpers.invoke(__ruby__.getCurrentContext(), this, "dispose");
        return (Object)ruby_result.toJava(Object.class);

    }

}

File: lib/ruby_console.rb
require 'jruby'
require 'irb'
require 'irb/completion'

#java_require 'ruby_console'  # Require this file, instead of pasting code into .java
java_package 'com.moneydance.modules.features.ruby.rb'

import java.awt.Color
import java.awt.Font
import java.awt.EventQueue
import javax.swing.JFrame
import javax.swing.JFileChooser
import com.moneydance.awt.AwtUtil
java_import 'java.awt.event.ActionListener' # Compiler needs this directive to implement ActionListener
java_import 'java.awt.event.ActionEvent' # Compiler needs this directive to implement ActionListener

# Moneydance IRB Console
# Implements ActionListener: The listener interface for receiving action events.
# Instance of class that implements this interface can be registered with any component
# using the component's #addActionListener method. When the action event occurs,
# registered listener's #actionPerformed method is invoked.
class RubyConsole
  java_implements ActionListener

  # Try to find preferred font family, use otherwise -- err -- otherwise
  def self.find_font otherwise, style, size, *preferred
    available = java.awt.GraphicsEnvironment.local_graphics_environment.available_font_family_names
    font_name = preferred.find { |name| available.include? name } || otherwise
    Font.new(font_name, style, size)
  end

  # Starts new RubyConsole
  def initialize ruby_main
    STDERR.puts "RubyConsole init called"
    @ruby_main = ruby_main

    if EventQueue.dispatch_thread?
      # Called from Moneydance GUI, use thread other than EventQueue.dispatch_thread
      Thread.new do
        run_irb
      end
    else
      # Called from command line (rake ant:load), we're not inside Moneydance GUI
      run_irb JFrame::EXIT_ON_CLOSE
    end
  end

  # Runs Swing wrapper for IRB - call from thread other than EventQueue.dispatch_thread
  #
  def run_irb close_operation = JFrame::HIDE_ON_CLOSE # DISPOSE_ON_CLOSE ?
    STDERR.puts "RubyConsole run_irb called"
    text = javax.swing.JTextPane.new
    text.font = RubyConsole.find_font 'Monospaced', Font::PLAIN, 14, 'Menlo', 'Monaco', 'Andale Mono'
    text.margin = java.awt.Insets.new(8, 8, 8, 8)
    text.caret_color = Color.new(0xa4, 0x00, 0x00)
    text.background = Color.new(0xf2, 0xf2, 0xf2)
    text.foreground = Color.new(0xa4, 0x00, 0x00)

    @irb_pane = javax.swing.JScrollPane.new
    @irb_pane.viewport_view = text

    @file_button = javax.swing.JButton.new "Load file"

    pane = javax.swing.JPanel.new java.awt.GridBagLayout.new
    pane.add(@irb_pane, AwtUtil.getConstraints(0, 0, 1, 1, 4, 1, true, true))
    pane.add(@file_button, AwtUtil.getConstraints(0, 3, 1, 0, 1, 1, false, true))

#    enableEvents(WindowEvent.WINDOW_CLOSING);
    @file_button.addActionListener(self)

    @frame = JFrame.new "Moneydance Interactive JRuby #{JRUBY_VERSION} Console " +
                            "(tab will autocomplete)"
    @frame.default_close_operation = close_operation
    @frame.set_size 800, 800
    @frame.content_pane.add pane # @frame.add pane

    header = " MD - Moneydance context: ComMoneydanceAppsMdController::Main \n" +
        " ROOT - Moneydance root account: ComMoneydanceAppsMdModel::RootAccount \n" +
        " TRANS - Moneydance TransactionSet: ComMoneydanceAppsMdModel::TransactionSet \n\n"

    readline = org.jruby.demo.TextAreaReadline.new text, header
    JRuby.objectspace = true # useful for code completion
    readline.hook_into_runtime_with_streams(JRuby.runtime)

    EventQueue.invoke_later proc { show }

    ARGV << '--readline' << '--prompt' << 'inf-ruby'
    IRB.start(__FILE__)
    @ruby_main.cleanup
  end

  java_signature 'public void actionPerformed(ActionEvent event)'
  # from ActionListener interface: Invoked when an action event occurs.
  def action_performed event
    case event.source
      when @file_button
        load_file
        show
      else
        STDERR.puts "Got unexpected event: #{event}"
    end
  end

  # Interactively loads Ruby script into IRB session
  #
  def load_file
    fc = JFileChooser.new
    fc.setDialogTitle("Choose Ruby Script File")
    if fc.showOpenDialog(@frame) == JFileChooser::APPROVE_OPTION
      @ruby_main.file fc.getSelectedFile.absolute_path
    else
      STDERR.puts "Unrecognized Ruby Script File"
    end
  end

  java_signature 'void show()'
  # Shows console window
  #
  def show
    @frame.set_visible true
    @frame.to_front
    @irb_pane.request_focus
  end

  java_signature 'void dispose()'
  # Properly disposes of Swing elements
  #
  def dispose
    @frame.set_visible false
    @frame.dispose
  end
end
